### PR TITLE
custom_matcher: Test the custom_match effectively

### PR DESCRIPTION
**Pull request by tyxia**
The existing test is using `NeverMatch` matcher which always return false for match() method. So, we didn't really test the matching behavior for custom_match field/mechanism. In this PR:
- Implement a new custom_matcher and factory to do simple string comparison matching.
- Test custom_match with this newly added custom matcher/factory for the expected matching result that `result.on_match_.has_value()` is true

Signed-off-by: Tianyu Xia <tyxia@google.com>

### CODE REVIEW

File: test/common/matcher/matcher_test.cc [hunk 1]
```diff
@@ -208,26 +208,31 @@ TEST_F(MatcherTest, CustomMatcher) {
...
 
   envoy::config::common::matcher::v3::Matcher matcher;
   MessageUtil::loadFromYaml(yaml, matcher, ProtobufMessage::getStrictValidationVisitor());
 
   TestUtility::validate(matcher);
 
-  auto inner_factory = TestDataInputBoolFactory("foo");
-  NeverMatchFactory match_factory;
+  // Build the input data that is to be matched.
+  std::string value = "custom_foo";
```
Comment (daixiang0) [inline test/common/matcher/matcher_test.cc:223]: Could make it const?

**Review by daixiang0 — APPROVED** [review]
LGTM

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**